File: src/Ebay/Server.Application.New/Abstractions/Queries/IProductQueries.cs
using Server.Application.New.Models;

namespace Server.Application.New.Abstractions.Queries;

/// <summary>
/// контракт.
/// </summary>
public interface IProductQueries
{
    /// <summary>
    /// Операция контракта приложения.
    /// </summary>
    Task<ProductInfo?> GetProductAsync(Guid productId, CancellationToken cancellationToken);

    //todo сделать асинхронную энумерацию или пагинацию
    /// <summary>
    /// Операция контракта приложения.
    /// </summary>
    Task<IReadOnlyList<ProductInfo>> GetAllProductsAsync(CancellationToken cancellationToken);

    /// <summary>
    /// Операция контракта приложения.
    /// </summary>
    Task<IReadOnlyList<Guid>> GetAllProductsIdsAsync(CancellationToken cancellationToken);
}

File: src/Ebay/Server.Application.New/Abstractions/Queries/IProductQueries.fs
namespace Server.Application.New.Abstractions.Queries

open System
open System.Collections.Generic
open System.Threading
open System.Threading.Tasks
open Server.Application.New.Models

/// <summary>
/// Порт чтения данных товаров.
/// </summary>
type IProductQueries =
    /// <summary>
    /// Возвращает карточку товара.
    /// </summary>
    /// <param name="productId">Идентификатор товара.</param>
    /// <param name="cancellationToken">Токен отмены операции.</param>
    /// <returns>Карточка товара либо <see langword="null" />, если товар не найден.</returns>
    abstract GetProductAsync: productId: Guid * cancellationToken: CancellationToken -> Task<ProductInfo>

    //todo сделать асинхронную энумерацию или пагинацию
    /// <summary>
    /// Возвращает список всех товаров.
    /// </summary>
    /// <param name="cancellationToken">Токен отмены операции.</param>
    /// <returns>Список товаров.</returns>
    abstract GetAllProductsAsync: cancellationToken: CancellationToken -> Task<IReadOnlyList<ProductInfo>>

    /// <summary>
    /// Возвращает идентификаторы всех товаров.
    /// </summary>
    /// <param name="cancellationToken">Токен отмены операции.</param>
    /// <returns>Список идентификаторов товаров.</returns>
    abstract GetAllProductsIdsAsync: cancellationToken: CancellationToken -> Task<IReadOnlyList<Guid>>
